### PR TITLE
Support encoding and decoding trust stores

### PR DESCRIPTION
**Pull request by lulf**

### CODE REVIEW

**Comment by lulf** [comment]
@AGWA Would you mind having a look at this PR?

File: pkcs12.go [hunk 1]
```diff
@@ -481,6 +529,81 @@ func Encode(rand io.Reader, privateKey interface{}, certificate *x509.Certificat
...
+func EncodeTrustStore(rand io.Reader, certs []*x509.Certificate, password string) (pfxData []byte, err error) {
+	encodedPassword, err := bmpString(password)
+	if err != nil {
+		return nil, err
+	}
+
+	var pfx pfxPdu
+	pfx.Version = 3
+
+	// This will make the certs usable in JDK >= 1.8
```
Comment (AGWA) [inline pkcs12.go:554]: Could this comment be made more specific - e.g. instead of "usable" say "trusted"

File: pkcs12.go [hunk 2]
```diff
@@ -481,6 +529,81 @@ func Encode(rand io.Reader, privateKey interface{}, certificate *x509.Certificat
...
+	var javaTrustStoreAttr pkcs12Attribute
+	javaTrustStoreAttr.Id = oidJavaTrustStore
+	javaTrustStoreAttr.Value.Class = 0
+	javaTrustStoreAttr.Value.Tag = 17
+	javaTrustStoreAttr.Value.IsCompound = true
+
+	var certBags []safeBag
+	var certBag *safeBag
```
Comment (AGWA) [inline pkcs12.go:562]: Please move `certBag` into the body of the for loop since it is not used elsewhere.

File: pkcs12.go [hunk 3]
```diff
@@ -301,7 +306,50 @@ func DecodeChain(pfxData []byte, password string) (privateKey interface{}, certi
 	return
 }
 
-func getSafeContents(p12Data, password []byte) (bags []safeBag, updatedPassword []byte, err error) {
+// DecodeChain extracts the certificates from a truststore.
```
Comment (AGWA) [inline pkcs12.go:309]: Please change "DecodeChain" to "DecodeTrustStore".

File: pkcs12.go [hunk 4]
```diff
@@ -207,6 +209,9 @@ func convertAttribute(attribute *pkcs12Attribute) (key, value string, err error)
 		// This key is chosen to match OpenSSL.
 		key = "Microsoft CSP Name"
 		isString = true
+	case attribute.Id.Equal(oidJavaTrustStore):
+		key = oidJavaTrustStore.String()
+		isString = true
```
Comment (AGWA) [inline pkcs12.go:214]: Considering `ToPEM` is buggy and deprecated, I would prefer not to add any new functionality to it.  Could you remove this change?

File: pkcs12.go [hunk 5]
```diff
@@ -301,7 +306,50 @@ func DecodeChain(pfxData []byte, password string) (privateKey interface{}, certi
...
+			certs = append(certs, parsedCerts[0])
+
+		default:
+			return nil, errors.New("pkcs12: expected only certificate bags")
+		}
+	}
+
+	if certs == nil {
+		return nil, errors.New("pkcs12: certificates missing")
+	}
```
Comment (AGWA) [inline pkcs12.go:347]: Is there a reason for this check? It doesn't seem like loading an empty trust store ought to be an error. The caller can always check themselves if they care.

Comment (lulf) [reply pkcs12.go]: I agree, I was in doubt whether to do this or not.

File: pkcs12.go (hunk 4 above)
Comment (lulf) [reply pkcs12.go]: Done, this is a leftover from my manual testing decoding keystores to check the attribute contents.

**Comment by lulf** [comment]
@AGWA I pushed a commit that addresses your review comments so far. Let me know if/when you want me to squash the commits.

**Comment by AGWA** [comment]
@luluf looks good, thanks! Please do go ahead and squash.

**Comment by lulf** [comment]
@AGWA Done, thank you for taking the time to review.

**Comment by AGWA** [comment]
Thank you for your contribution, @lulf!  BTW, I just realized that `DecodeTrustStore` ought to be checking for the attribute so it can take advantage of the safety feature too, so I quickly coded that up (2981468c0ff3c42cbfd8601483b9e16fea379187)

**Pull request merged**